src/html.rs

```diff
@@ -92,36 +152,184 @@ fn remove_ids(document: &mut Html, ids: impl IntoIterator<Item = NodeId>) {
     }
 }
 
-fn is_empty_or_whitespace(el: &ElementRef) -> bool {
-    el.text().flat_map(str::chars).all(char::is_whitespace)
+fn remove_non_element_nodes(document: &mut Html) {
+    let mut to_remove = Vec::new();
+    // `.root_element()` returns the first `Element` in the children of the
+    // root, which comments/doctypes are not.
+    // Use `root()` instead and `skip()` because `descendants` includes the
+    // node it is called on.
+    for el in document.tree.root().descendants().skip(1) {
+        if el.value().is_comment() || el.value().is_doctype() {
+            to_remove.push(el.id());
+        }
+    }
+    remove_ids(document, to_remove.drain(..));
+}
+
+fn remove_toplevel_whitespace(document: &mut Html) {
+    let mut to_remove = Vec::new();
+
+    let parent = document.tree.root();
+
+    for el in parent.children() {
+        let Some(t) = el.value().as_text() else {
+            continue;
+        };
+
+        if t.chars().all(char::is_whitespace) {
+            to_remove.push(el.id());
+        }
+    }
+
+    trace!(
+        "Removing {} whitespace text nodes children from {:?}",
+        to_remove.len(),
+        parent.value(),
+    );
+    remove_ids(document, to_remove.drain(..));
+}
+
+fn remove_empty(document: &mut Html) {
+    let mut to_remove = Vec::new();
+
+    for el in document
+        .root_element()
+        .descendants()
+        .filter_map(ElementRef::wrap)
+    {
+        if is_empty_or_whitespace(&el) {
+            to_remove.push(el.id());
+        }
+    }
+
+    remove_ids(document, to_remove.drain(..));
+}
+
+fn remove_empty_sections(document: &mut Html) {
+    let mut to_remove = Vec::new();
+    for el in document.select(&HEADERS) {
+        // TODO: does select match on detached nodes?
```

Comment:
Is it hard to test?

Reply:
I just checked that for 5835ae4, [it doesn't](https://docs.rs/scraper/latest/src/scraper/element_ref/mod.rs.html#90) and I've removed the comment.